src/driver/linux_onload/tcp_sendpage.c

```diff
@@ -54,7 +54,6 @@ ssize_t linux_tcp_helper_fop_sendpage(struct file* filp, struct page* page,
   ci_assert(page);
   ci_assert_ge(offset, 0);
   ci_assert_gt(size, 0);
-  ci_assert_le(offset + size, CI_PAGE_SIZE);
```

Comment:
Would it be helpful to keep the assert if the page is huge/THP? e.g.
```
ci_assert((offset + size <= CI_PAGE_SIZE) || PageCompound(page) || PageHuge(page));
```

Reply:
No.  `PageCompound(page)` is true for the first "small" page of the compound; it is false for all the following pages.

Reply:
> `PageCompound(page)` is true for the first "small" page of the compound; it is false for all the following pages.

Huh, that's interesting! I guess it's a newer kernel feature.

Could you try running https://gist.github.com/ivatet-amd/68a808e2e571871f211c145157fc8825? In my older kernels, it shows the THP and HUGE bits set across the entire huge or TH page:
```
$ echo 10 | sudo tee /proc/sys/vm/nr_hugepages
$ sudo ./test.elf
THP
pagemap: 0x81800000000c6400 kpageflags: 0x40d868 HEAD THP
pagemap: 0x81800000000c6401 kpageflags: 0x411800 TAIL THP
pagemap: 0x81800000000c6402 kpageflags: 0x411800 TAIL THP
pagemap: 0x81800000000c6403 kpageflags: 0x411800 TAIL THP
Hugepage
pagemap: 0x818000000009d000 kpageflags: 0x29818 HEAD HUGE
pagemap: 0x818000000009d001 kpageflags: 0x800031800 TAIL HUGE
pagemap: 0x818000000009d002 kpageflags: 0x31800 TAIL HUGE
pagemap: 0x818000000009d003 kpageflags: 0x31800 TAIL HUGE
```

I believe `PageCompound()` would have returned "true" in the kernel space. Does it return "false" and break down the assertion?

Reply:
>  I'm unable to reproduce this issue now.

That's fine. My point was about relaxing the assertion rather than dropping it entirely.

Would it work for you if (1) we merge the suggested patch, and (2) if the repro happens again with the updated assertion, we drop it?

Reply:
I believe that you run Socket Tester, and I guess it happens to show the issue from time to time.  So it seems reasonable to apply a modified patch and look at the test results.

Reply:
I've checked with @rhughes-xilinx, and we agreed to drop the assertion, considering `sendpage` is dropped from FOPs in the latest kernel/has a limited lifespan.